### ISSUE TITLE
Better segment hovering raycast

### ISSUE THREAD

**Issue by kianzarrin** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#576]
in #564 I made it easier to hover over small segment. Even so hovering can still be hard for small junctions. I think this is a caveat with skylines built-in ray-cast but I think I have a work around.

Here is the problem:
![Cities_Ct6sactalI](https://user-images.githubusercontent.com/26344691/69778384-c677ab80-11ac-11ea-8a0f-b893edc30b1e.png)
In the picture above I have demonstrated a situation in which its very hard to select the highlighted segment end. In fact the 2 mouse positions in the picture above are the only places the mouse can be to select the desired segment end. You notice that the mouse is near the right segment but still the left segment is selected. This is not right. I think that the selection areas should be like this:
![Screenshot (124)](https://user-images.githubusercontent.com/26344691/69778571-75b48280-11ad-11ea-829a-66a0940f2d2e.png)
Terminology:
- hover area: the area the mouse pointer should be in in order to chose the desired segment end. In picture above its red stakes.
- highlight shape: the shape of the highlight which is a blue half sausage.

Note that I want to change the hover area only. The highlight shape is and shall remain half sausage.

**Comment by kianzarrin** [comment]
To solve this issue I suggest to do this with vector directions
![Screenshot (123) - Copy](https://user-images.githubusercontent.com/26344691/69778892-cf697c80-11ae-11ea-8372-88ed8535422b.png)
in picture above :
 - green arrows: end direction vectors of the segments.
 - white triangle: mouse pointer
 - red arrow: direction vector connecting the mouse pointer to the node.
 - yellow arc: angle between the mouse vector and the segment vector.

So once we find the hovered node ID then we can iterate through its segments. The desired segment is the one that makes the minimum angle with the mouse vector.

The code bellow shows the concept. Though I think it fails to calculate the direction vector between mouse and node properly.
```C#
private ushort GetSegmentFromNode() {
    ushort minSegId = 0;
    if (HoveredNodeId != 0) {
        var netService = Constants.ServiceFactory.NetService;
        NetNode[] nodes = Singleton<NetManager>.instance.m_nodes.m_buffer;
        //TODO does this actually find the mouse direction vector?
        var dir0 = m_mouseRay.origin - nodes[HoveredNodeId].m_position;
        float min_angle = float.MaxValue;

        netService.IterateNodeSegments(
            HoveredNodeId,
            (ushort segmentId, ref NetSegment segment) => {
                Vector3 dir;
                if (segment.m_startNode == HoveredNodeId) {
                    dir = segment.m_startDirection;
                } else {
                    dir = segment.m_endDirection;
                }
                float angle = Vector3.AngleBetween(dir0, dir);
                if (angle < min_angle) {
                    min_angle = angle;
                    minSegId = segmentId;
                }
                return true;
            });
    }
    return minSegId;
}
```

**Comment by aubergine10** [comment]
This would be great when camera is looking downwards, but what happens if user has camera tilted upwards (looking more towards horizon) and tries to select a segment on the far side of a junction?

**Comment by kianzarrin** [comment]
I am haven't done casting before but if we could cast the mouse on the train and find the coordinates that would be great.

**Comment by kvakvs** [comment]
The terrain can be uneven, streets will approach the junction from all 3d directions. So yes, cast to the terrain, and then solve in 3d space.

**Comment by aubergine10** [comment]
Note, however, that _transition_ and _junction_ nodes always flatten terrain, regardless of existing terrain or slope of segments connected to them. (It's why doing junctions on sloped rail tracks is such a nightmare).

**Comment by aubergine10** [comment]
Random idea: Could we buff the clickability of segments based on their length? So the shortest segment is easier to select via the node (it's 'select triangle' within the node would be enlarged slightly)?

**Comment by kianzarrin** [comment]
> Random idea: Could we buff the clickability of segments based on their length? So the shortest segment is easier to select via the node (it's 'select triangle' within the node would be enlarged slightly)?

**Comment by kianzarrin** [comment]
I got this working on my test tool: https://github.com/kianzarrin/KianHoverTool/blob/450a5bf42b43e4361b2f9d5b4530780b21b18022/KianHoverElements/Tool/KianToolBase.cs#L169
I have not tried "buff the clickability of segments based on their length"
I have not taken vertical angle into consideration I set the vertical axes to 0.

**Comment by kianzarrin** [comment]
I integrated it into `DetermineHoveredElements()` code from TMPE.

@aubergine10  I managed to "buff the clickability of segments based on their length" as you suggested. easy peasy fun! But do you think I should do it? I will make it depend on a boolean so that we can switch forth and back at will.
pro: slightly easier to select small segments (small advantage) 
con: sophistication may confuse user (small issue).

**Comment by kianzarrin** [comment]
I tested this on roads with bridges over. It works no-problem!

**Comment by kianzarrin** [comment]
I tested this with trains. No improvements with regards to level crossings. I will look into this.

**Comment by kianzarrin** [comment]
The reason why it does not work on level crossing is that this code does not find nodes of level crossing:
```c#
                    // find train nodes
                    nodeInput.m_netService.m_itemLayers =
                        ItemClass.Layer.Default | ItemClass.Layer.MetroTunnels ;
                    nodeInput.m_netService.m_service = ItemClass.Service.PublicTransport;
                    nodeInput.m_netService.m_subService = ItemClass.SubService.PublicTransportTrain;
                    nodeInput.m_ignoreNodeFlags = NetNode.Flags.None;
                    // nodeInput.m_ignoreNodeFlags = NetNode.Flags.Untouchable;

                    if (RayCast(nodeInput, out nodeOutput))
                    {
```

I did not manage to hover level crossing nodes. TMPE hover such nodes by first finding the hovered segment and then looking at segment ends. It feels weird to find the node based on segment and then find segment based on node. but its worth a shot.

To

**Comment by kianzarrin** [comment]
my pull request #615 is merged in but I am having second thoughts about it! In particular this:https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/pull/615/files/ced446c5fbf2ca559049df904f7eb56f25479d03#r368252980

I am looking at segment end to mouse pointer angle even when the mouse pointer is not at node but rather in the middle of a segment. Now that I think about it maybe it is possible to bend the roads so extremely to make the wrong segment selected. ~~I tried but did not managed to find a way to make the code fail. does not mean that it can't ~~ I manged to reproduce the problem using move it mod:
![Screenshot (343)](https://user-images.githubusercontent.com/26344691/73133795-61b6c100-4036-11ea-949c-98c928366efa.png)
I can call `GetHoveredSegmentFromNode()` earlier but that way Level crossings will not benefit from the node hovering boost. 
I can also get rid of the angle based code and use `GetClosestPositionAndDirection()`
